### PR TITLE
An EJS template to populate the TLA+ spec of xycloans

### PR DESCRIPTION
**Pull request by konnov**
This PR contains two contributions in the context of the xycloans case study:
 
 - [x] a custom EJS template to translate an aggregated state of the xycloans contract to TLA+. I've managed to generate several transactions with Apalache.
 - [x] a simple script to ingest the aggregated state and a counterexample to produce a Stellar CLI command.

We still have to figure out a reasonable UX.

Ignore the commits below [3b3cb3e](https://github.com/freespek/solarkraft/pull/156/commits/3b3cb3e597d6e8c259c64e9ba7c16594ee90876f), as they were propagated via `git merge`.

### CODE REVIEW

File: doc/case-studies/xycloans/ingest.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env node
+/**
+ * Injest a counterexample produced by Apalache and produce the corresponding
```
Comment (thpani) [inline doc/case-studies/xycloans/ingest.js:3]:
```suggestion
 * Ingest a counterexample produced by Apalache and produce the corresponding
```

File: doc/case-studies/xycloans/ingest.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * command for stellar-cli.
+ *
+ * In this case study, we write the script manually.  In the future, this could
+ * be automated. Alternatively, we could execute the counterexample with
+ * stellar-sdk. However, we believe that the command line interface offers us
```
Comment (thpani) [inline doc/case-studies/xycloans/ingest.js:8]:
```suggestion
 * js-stellar-sdk. However, we believe that the command line interface offers us
```

File: doc/case-studies/xycloans/ingest.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+if (args.length < 2) {
+  console.log('Usage: ingest.js state.json trace.json')
+  console.log('  state.json is the aggregated state, as produced by solarkraft aggregate')
+  console.log('  trace.json is the ITF trace, as produced by Apalache')
+  process.exit(1)
+}
+
+// read the state and the trace from the JSON files
+const state = JSON.parse(fs.readFileSync(args[0], 'utf8'))
+const trace = JSON.parse(fs.readFileSync(args[1], 'utf8'))
```
Comment (thpani) [inline doc/case-studies/xycloans/ingest.js:30]: Do we need a try/catch around this? :)

File: doc/case-studies/xycloans/MCxycloans_monitor.ejs.tla [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    <%
+    const balanceAddrs =
+        Object.keys(storage[contractId].persistent)
+            .filter((key) => key.startsWith("Balance,"))
+            .map((key) => key.split(",")[1])
+    %>
+    <%-
+    balanceAddrs
+        .map((addr) => `    "${addr}"`)
+        .join(",\n    ")
```
Comment (thpani) [inline doc/case-studies/xycloans/MCxycloans_monitor.ejs.tla:34]: Bit confused about the whitespace here. Looks like you're indenting twice?

File: doc/case-studies/xycloans/MCxycloans_monitor.ejs.tla [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    const tokenAddrs =
+        Object.keys(storage[storage[contractId].instance.TokenId].persistent)
+            .filter((key) => key.startsWith("Balance,"))
+            .filter((key) => key !== `Balance,${contractId}`)
+            .map((key) => key.split(",")[1])
+    %>
+    <%-
+    tokenAddrs
+        .map((addr) => `    "${addr}"`)
+        .join(",\n    ")
```
Comment (thpani) [inline doc/case-studies/xycloans/MCxycloans_monitor.ejs.tla:50]: Bit confused about the whitespace here. Looks like you're indenting twice?

File: doc/case-studies/xycloans/MCxycloans_monitor.ejs.tla [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            %>})
+        ],
+        token_persistent |-> [ Balance |-> SetAsFun({<%-
+              Object.keys(storage[storage[contractId].instance.TokenId].persistent)
+                .filter((key) => key.startsWith("Balance,"))
+                .map((key) => key.split(",")[1])
+                .map((addr) =>
+                  `<<"${addr}", ${storage[storage[contractId].instance.TokenId].persistent["Balance," + addr].amount}>>`)
+                .join(",\n              ")
+        %>})]
```
Comment (thpani) [inline doc/case-studies/xycloans/MCxycloans_monitor.ejs.tla:109]: Can we extract this into some kind of helper?
It looks like we need this pattern again and again.

**Review by thpani — APPROVED** [review]
I like the template approach better than I thought 👍

File: doc/case-studies/xycloans/MCxycloans_monitor.ejs.tla (hunk 4 above)
Comment (konnov) [reply doc/case-studies/xycloans/MCxycloans_monitor.ejs.tla:34]: oh, that's true. Copilot is doing weird things sometimes :)

File: doc/case-studies/xycloans/MCxycloans_monitor.ejs.tla (hunk 6 above)
Comment (konnov) [reply doc/case-studies/xycloans/MCxycloans_monitor.ejs.tla:109]: Good point. I have found a way to introduce definitions in EJS

File: doc/case-studies/xycloans/ingest.js (hunk 3 above)
Comment (konnov) [reply doc/case-studies/xycloans/ingest.js:30]: yep, just added

**Pull request merged**